Validate `PageProps` route literal against known routes (only fall back to `string` when there are none)

## Summary

`PageProps<Path>` (from `waku/router`) doesn't validate `Path` against the routes the type system already knows about. A non-existent or typo'd route literal such as `PageProps<'/blog/[wrong]'>` is accepted and resolves to a structurally-derived props type instead of producing a type error. The slug name is taken from whatever is inside the brackets, so the mistake only surfaces later (when you destructure the prop name you *expected*), or not at all if the names happen to line up.

## Current behavior

Definition (`waku/dist/router/base-types.d.ts`):

```ts
export type PageProps<Path extends PagePath<CreatePagesConfig> | (string & {})> =
  PropsForPages<Path>;
```

Because the constraint includes `(string & {})`, any string satisfies it, so `Path` is never checked against `PagePath<CreatePagesConfig>` (the routes contributed by the generated `pages.gen.ts`).

Resolved type of a bogus route. It is **not** `never` (an `[X] extends [never]` probe returns `false`), and the `wrong` slug is extracted structurally from the literal:

```
PageProps<'/blog/[wrong]'>
// => { path: `/blog/${string}`; query: string; wrong: string }
```

In a page that actually lives at `src/pages/blog/[slug].tsx`:

```tsx
import type { PageProps } from 'waku/router';

export default function BlogPage({ slug }: PageProps<'/blog/[wrong]'>) {
  return <h1>{slug}</h1>;
}
```

`tsc` only complains downstream about the destructure, not about the wrong route:

```
error TS2339: Property 'slug' does not exist on type
  '{ path: `/blog/${string}`; query: string; wrong: string; }'.
```

If the wrong name happened to match a real slug, nothing would be flagged at all.

## Why this happens

The `(string & {})` arm is presumably there so `PageProps` stays usable **before** `pages.gen.ts` is generated: when `CreatePagesConfig` is empty, `PagePath<CreatePagesConfig>` is `never`, and without a fallback every `PageProps<...>` would error on the constraint. The side effect is that the fallback is *always* active (even after codegen), which disables route validation entirely.

## Proposed change

Only widen to `string` when there are no known routes (i.e. `PagePath<CreatePagesConfig>` is `never`):

```ts
export type PageProps<
  Path extends [PagePath<CreatePagesConfig>] extends [never]
    ? string
    : PagePath<CreatePagesConfig>,
> = PropsForPages<Path>;
```

The tuple wrapping (`[...] extends [never]`) is load-bearing so the `never` check is non-distributive.

## Experimental results

### Isolated repro (mirrors the real type shape)

**Empty config (pre-codegen) - any string still accepted:**

```ts
interface CreatePagesConfig {}
type PagePath<C> = C extends { pages: { path: infer P } } ? P : never;
type GetSlug<P> = P extends `${string}[${infer S}]${string}` ? S : never;
type SlugTypes<P extends string> = { [K in GetSlug<P>]: string };
type PropsForPages<P extends string> = { path: P } & SlugTypes<P>;

type Constraint = [PagePath<CreatePagesConfig>] extends [never] ? (string & {}) : PagePath<CreatePagesConfig>;
type PageProps<Path extends Constraint> = PropsForPages<Path>;

type T = PageProps<'/literally/anything'>; // expect OK
```

```
$ tsc --strict --noEmit caseA.ts
NO ERRORS (string fallback works)
```

**Populated config (post-codegen) - unknown route rejected, valid route fine:**

```ts
interface CreatePagesConfig { pages: { path: '/' } | { path: '/blog/[slug]' }; }
// ...same machinery...
type Ok  = PageProps<'/blog/[slug]'>;   // expect OK
type Bad = PageProps<'/blog/[wrong]'>;  // expect error
```

```
$ tsc --strict --noEmit caseB.ts
caseB.ts: error TS2344: Type '"/blog/[wrong]"' does not satisfy the constraint '"/" | "/blog/[slug]"'.
```

### Against the real types (patched `base-types.d.ts`, real app + generated `pages.gen.ts`)

The app has `src/pages/index.tsx` and `src/pages/blog/[slug].tsx`, so the generated `pages.gen.ts` makes `PagePath<CreatePagesConfig> = '/' | '/blog/[slug]'`.

**Valid route - clean:**

```tsx
export default function BlogPage({ slug }: PageProps<'/blog/[slug]'>) { /* ... */ }
```
```
$ tsc --noEmit
# (clean)
```

**Non-existent route - now errors at the type argument:**

```tsx
export default function BlogPage({ slug }: PageProps<'/blog/[wrong]'>) { /* ... */ }
```
```
error TS2344: Type '"/blog/[wrong]"' does not satisfy the constraint '"/" | "/blog/[slug]"'.
error TS2339: Property 'slug' does not exist on type
  '{ path: `/blog/${string}`; query: string; wrong: string; }'.
```

The pre-codegen behavior is unchanged (empty config still accepts any string), so there's no regression for the bootstrap scenario the fallback was protecting.

## Notes / open questions

- **Autocomplete is preserved.** A union constraint still offers its members as completions for the type argument, *and* now rejects invalid literals, strictly more than `(string & {})` provided.
- **Edge case to validate before merging.** If any page's `path` ever infers as a non-literal `string` (rather than a `'/...'` literal, e.g. certain catch-all (`[...slug]`), nested multi-slug (`[a]/[b]`), or `getConfig`-inference cases), then `PagePath<CreatePagesConfig>` would include `string`, collapsing the constraint back to "accept anything." That's graceful (same as the empty-config case, no error), but it would silently turn validation off for that route set. Only single-slug `[slug]` and a static `/` were tested here; catch-all / nested / optional shapes still need a check.

## Environment

- `waku@1.0.0-beta.3` (please confirm against latest)
- TypeScript `6.0.x`, `--strict`
- Verified with `tsc --noEmit`

--- a/e2e/fixtures/fs-router/src/page-props-typing.ts
+++ b/e2e/fixtures/fs-router/src/page-props-typing.ts
@@ -2,16 +2,25 @@ import type { PageProps } from 'waku/router';
 
 // Type-level assertions only; never called. CreatePagesConfig.pages is augmented
 // by the generated pages.gen.ts, so a known route literal resolves to typed
-// props while an unknown one is rejected at the type parameter.
+// props while an unknown one is rejected at the type parameter. fs-router emits
+// every path as a string literal (including catch-all and nested segments), so
+// the union never widens to `string` and validation stays active for all shapes.
 export function assertPagePropsTyping(
   staticProps: PageProps<'/bar'>,
   dynamicProps: PageProps<'/nested/[name]'>,
+  nestedSegmentProps: PageProps<'/page-with-segment/article/[slug]'>,
+  catchAllProps: PageProps<'/subroute/[...catchAll]'>,
 ) {
   void staticProps.path;
   void dynamicProps.name;
+  void nestedSegmentProps.slug;
+  const segments: string[] = catchAllProps.catchAll;
+  void segments;
 }
 
 // @ts-expect-error an unknown route literal is not a known page path
 export type InvalidPageProps = PageProps<'/no-such-route'>;
 // @ts-expect-error a wrong slug name does not match the known dynamic route
 export type WrongSlugPageProps = PageProps<'/nested/[wrong]'>;
+// @ts-expect-error a wrong catch-all name is rejected even with a catch-all route present
+export type WrongCatchAllProps = PageProps<'/subroute/[...wrong]'>;

--- a/e2e/fixtures/fs-router/src/page-props-typing.ts
+++ b/e2e/fixtures/fs-router/src/page-props-typing.ts
@@ -1,0 +1,17 @@
+import type { PageProps } from 'waku/router';
+
+// Type-level assertions only; never called. CreatePagesConfig.pages is augmented
+// by the generated pages.gen.ts, so a known route literal resolves to typed
+// props while an unknown one is rejected at the type parameter.
+export function assertPagePropsTyping(
+  staticProps: PageProps<'/bar'>,
+  dynamicProps: PageProps<'/nested/[name]'>,
+) {
+  void staticProps.path;
+  void dynamicProps.name;
+}
+
+// @ts-expect-error an unknown route literal is not a known page path
+export type InvalidPageProps = PageProps<'/no-such-route'>;
+// @ts-expect-error a wrong slug name does not match the known dynamic route
+export type WrongSlugPageProps = PageProps<'/nested/[wrong]'>;

--- a/packages/waku/src/router/base-types.ts
+++ b/packages/waku/src/router/base-types.ts
@@ -27,5 +27,7 @@ export interface SearchCodecsConfig {
 
 /** Props for pages when using `createPages` */
 export type PageProps<
-  Path extends PagePath<CreatePagesConfig> | (string & {}),
+  Path extends [PagePath<CreatePagesConfig>] extends [never]
+    ? string
+    : PagePath<CreatePagesConfig>,
 > = PropsForPages<Path>;

--- a/packages/waku/tests/page-props-types.test.ts
+++ b/packages/waku/tests/page-props-types.test.ts
@@ -13,6 +13,7 @@ describe('PageProps route constraint', () => {
     expectType<TypeEqual<Props['foo'], string>>(true);
     expectType<TypeEqual<Props['path'], `/made/up/${string}`>>(true);
     expectType<TypeEqual<Props['query'], string>>(true);
-    expect(true).toBe(true);
+    const props: Props = { path: '/made/up/x', query: '', foo: 'x' };
+    expect(props.foo).toBe('x');
   });
 });

--- a/packages/waku/tests/page-props-types.test.ts
+++ b/packages/waku/tests/page-props-types.test.ts
@@ -1,0 +1,18 @@
+import { expectType } from 'ts-expect';
+import type { TypeEqual } from 'ts-expect';
+import { describe, expect, it } from 'vitest';
+import type { PageProps } from '../src/router/base-types.js';
+
+// In this suite CreatePagesConfig is not augmented, so PagePath resolves to
+// never and PageProps falls back to accepting any string (the pre-codegen
+// bootstrap behavior). Rejection of unknown routes after codegen is proven in
+// the augmented fs-router fixture (page-props-typing.ts).
+describe('PageProps route constraint', () => {
+  it('accepts any string before codegen and types its props', () => {
+    type Props = PageProps<'/made/up/[foo]'>;
+    expectType<TypeEqual<Props['foo'], string>>(true);
+    expectType<TypeEqual<Props['path'], `/made/up/${string}`>>(true);
+    expectType<TypeEqual<Props['query'], string>>(true);
+    expect(true).toBe(true);
+  });
+});

